xbmc/video/dialogs/GUIDialogVideoVersion.cpp

```diff
@@ -0,0 +1,923 @@
+      int idVideoVersion = m_database.AddVideoVersionType(strVideoVersion, VideoVersionTypeOwner::AUTO);
+
+      m_database.AddExtrasVideoVersion(itemType, dbId, idVideoVersion, &item);
+
+      // get updated video versions
+      m_extrasVideoVersionList->Clear();
+      m_database.GetVideoVersions(itemType, dbId, *m_extrasVideoVersionList, VideoVersionItemType::EXTRAS);
+      m_extrasVideoVersionList->SetContent(CMediaTypes::ToPlural(type));
+
+      // refresh the list
+      CGUIMessage msg(GUI_MSG_LABEL_BIND, GetID(), CONTROL_LIST_EXTRAS_VIDEOVERSION, 0, 0, m_extrasVideoVersionList);
+      OnMessage(msg);
+    }
+  }
+}
+
+int CGUIDialogVideoVersion::NewVideoVersion(std::string& strVideoVersion)
+{
+  // prompt for the new video version
+  if (!CGUIKeyboardFactory::ShowAndGetInput(strVideoVersion, CVariant{g_localizeStrings.Get(40004)}, false))
+    return -1;
+
+  CVideoDatabase videodb;
+  if (!videodb.Open())
+    return -1;
+
+  return videodb.AddVideoVersionType(StringUtils::Trim(strVideoVersion), VideoVersionTypeOwner::USER);
+}
+
+bool CGUIDialogVideoVersion::SetVideoVersion(const CFileItem* item)
+{
+  // set this video as an additional version of another video, user will
+  // be prompted to provide the target video and the video version, and
+  // this video entry will be deleted from DB
+  //
+  // N.B. Choosing the video itself as target video is also allowed, which
+  //      will just change the video version type
+
+  if (!item || !item->HasVideoInfoTag())
+    return false;
+
+  const std::string& type = item->GetVideoInfoTag()->m_type;
+  VideoDbContentType itemType = item->GetVideoContentType();
+
+  std::string dbType;
+  if (itemType == VideoDbContentType::MOVIES)
+  {
+    dbType = "movies";
+  }
+  else
+    return false;
+
+  CVideoDatabase videodb;
+  if (!videodb.Open())
+    return false;
+
+  // invalid operation warning
+  CFileItemList list;
+  int dbId = item->GetVideoInfoTag()->m_iDbId;
+
+  videodb.GetVideoVersions(itemType, dbId, list);
+  if (list.Size() > 1)
+  {
+    CGUIDialogOK* dialog = CServiceBroker::GetGUI()->GetWindowManager().GetWindow<CGUIDialogOK>(WINDOW_DIALOG_OK);
+    if (!dialog)
+      return false;
+
+    dialog->SetHeading(CVariant{40005});
+    dialog->SetLine(1, StringUtils::Format(g_localizeStrings.Get(40006), CMediaTypes::GetLocalization(type)));
+    dialog->Open();
+
+    return false;
+  }
+
+  list.ClearItems();
+
+  // choose the target video
+  std::string videoTitleDir = StringUtils::Format("videodb://{}/titles", dbType);
+  if (itemType == VideoDbContentType::MOVIES)
+    videodb.GetMoviesNav(videoTitleDir, list);
+  else
+    return false;
+
+  if (list.Size() < 1)
+    return false;
+
+  list.Sort(SortByLabel, SortOrderAscending, CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(CSettings::SETTING_FILELISTS_IGNORETHEWHENSORTING) ? SortAttributeIgnoreArticle : SortAttributeNone);
+
+  int itemId = 0;
+  for (int i = 0; i < list.Size(); i++)
```

Comment:
Don't like ++i. And the example code shows my style :) https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#31-braces

Reply:
> Don't like ++i. 

 That's not a valid point. prefix op is always to prefer due to performance considerations
 
 Regarding being mentioned in our coding guidelines: Don't be selfish, it's in a section to define braces, so "unrelated". Ofc. this should be fixed not to include things we do not want to see in our code base. Thanks for the hint.

Reply:
Using ++i is for performance reason, are you kidding, right? I'm working on assembly language level. How can the same INC instruction have different performance? I'm trying to follow Kodi coding guidelines and there is nothing mentioned to use ++i but not i++.

Reply:
was curious on this one myself, while ++i and i++ do have a difference when the variable is being used in other things as when you increment matters.. which does not apply to for loops. looking I found this which explains that ++i is technically more efficient:

https://stackoverflow.com/questions/24886/is-there-a-performance-difference-between-i-and-i-in-c
> the question one should be asking is not which of these two operations is faster, it is which of these two operations expresses more accurately what you are trying to accomplish. I submit that if you are not using the value of the expression, there is never a reason to use i++ instead of ++i, because there is never a reason to copy the value of a variable, increment the variable, and then throw the copy away.

Reply:
@xodidox 

> Using ++i is for performance reason, are you kidding, right? I'm working on assembly language level. How can the same INC instruction have different performance?

Short answer: it doesn't have different performance, and the question is generally moot outside of tight loops.

Team answer: We care more for code intention. ++ prefix implies that you wrote the code without the side effects of ++ postfix. Prefix vs. postfix implies the intention of the code, and in for loops the intention is clearly prefix.

To answer @thezoggy , optimal efficiency outside of a hard loop is something we rarely look at.

Reply:
> How can the same INC instruction have different performance?

 Because at C++ level there is much more code involved than just the ASM INC instruction.
 
As for example Scott Meyers (and others) have explained, for the implementation of the postfix operator, you need to first take a copy of the old value, because you need to return it, then increment the value. The prefix operator can be implemented without making a copy, because it returns the new value. But yeah, there are rumors that modern compilers are optimizing this away. I personally never would write code that relies on the optimization capabilities of a compiler, because afaik those are not standardized and are not mandatory to be implemented by compilers (but I'm not c++ standard guru and might be wrong here).

Reply:
> Compiler won't generate anything except a single INC instruction for this.

Proof, that there is a specification that *every* compiler MUST implement this optimization. I'm not aware of any (but might be wrong).

> the return value of the expression is ignored, so they are exactly same.

And this is exactly were we are heading for. It's actually not performance, let's put this aside for now. It's about intention, to make clear that you by intention will not use the return value.

> I'm old enough

Let's try to also put those kind of personal things aside, please. This kind of statements lead to nowhere.

Reply:
It's good to see you agree that it's not perf issue. But I didn't see any requirements in the guidelines for this after today's guidelines update. Like I mentioned before, if it's mandatory, document it, I'll stick with it; if not, I'll use what I like.

Reply:
Are you really like that? You will stick to whatever we put in the guideline, but if we just ask you to do it, you deny? That’s just crazy, but so be it. I will update the guideline today.

Reply:
Updated. It's not crazy at all from my point of view. For any rules that need to be followed, I want them to be documented and clear enough.

Reply:
Mate, you're trying to contribute to a project. Chances are more than likely you'll do one contribution and never again. That means the "team" has the long term maintenance burden. If a team member requests a change, just do it, its not that hard, and surely by now changing this as requested instead of 2 other prs, and how many more "but it doesn't explicitly say this" conversations are going to stem from this attitude is just a pain in the ass for all involved

Reply:
@fuzzard  I don't think you see the problem here. You did mention "team", right? Does a team only have one member? If not, which one's opinion should I follow? Especially for things like this, which has no performance difference, and everyone has their own favor? The answer is no one.
I should follow the agreement between team members, that's the document. @ksooo mentioned ++i, the doc itself uses i++, there is already conflict, nerveless to say i++ is used in the code extensively. Now the doc is updated, the Kodi team agrees on the ++i, I can follow that easily.

Reply:
Now you're just starting bullshit. At no point has any team member disagreed about the change requested by ksooo. If you need us to explicitly detail in a document every aspect before you respect a change request from a team member, fuck off, and don't waste our time.

Reply:
Watch your mouth! Who is starting bullshit? I read multiple times of the coding guidelines before and now I got a review opinion which is conflicting with the sample code in the guideline, isn't that a disagreement? Aren't the coding guidelines writing by a team member? Who is wasting time? When did I ask for documenting every aspect of details? I only asked for document this conflict!

Reply:
Whoa whoa, calm down. @xodidox please recognize that you're working on an open source contribution. In open source we value the heterogeneity of opinions, and I'm glad that our members feel comfortable in voicing theirs even if it's contrary in some way. The last thing I want is for anyone to feel attacked for their opinion, which it seems some of my team members do.

You raise a fair point: outsiders need guidance via documentation. One wonderful thing about our project is the effort our members have put into documentation, all the while recognizing that it's a living document defined by the edge cases we get in contributions like yours. We've recognized the deficiencies in our docs (which there always will be) and worked to updated them for your benefit. Please respect that.

Be aware that as patience wears thin, this PR risks rejection. Team Kodi is a bit different in that we've been drinking beers together for decades. If a single one of my team members feels harmed by your comment and closes the PR the entire team will stand by them 100%.

Reply:
Lets push this PR to the merged state and drink some beer then, no? :)

Reply:
To merged it is. And good fucking luck beating @phunkyfish and me 😉

Reply:
As someone looking at this from the outside (subscribed because of the potential utility for jellyfin-kodi), the whole argument over `i++` vs `++i` looks silly (no offense intended).

At best, it looks like an argument over runtime optimization, with no performance tests to back up either side of the argument. At worst, it looks like an argument over code formatting, with no linter configuration to back it up.

I can't comment on the implementation, as C++ isn't my forte, and I'm not intimately familiar with the Kodi codebase (I know it mostly from the Python side).

This feature looks desirable to me, as it would bring a feature to Kodi that has been available in streaming servers such as Plex, Emby and Jellyfin for a long time. It would potentially allow better integration between Kodi and media servers (I've yet to look at what exactly would be required on my end for this).

@xodidox has been working on this feature on and off for *5 years*, resolving merge conflicts along the way, and addressing an absolute massive wall of feedback over multiple iterations, putting in many hours of work.
If the team feels strongly about `i++` vs `++i` (which looks to me like what could easily be considered an optional nitpick), the path of least resistance could in cases such as this be to simply accept the PR, and then after it has been merged, follow up with short optimization PR.
Dealing with PR comments/change requests which seem inconsequential can be exhausting, especially when there are many of them over several rounds of reviews.

[![XKCD Optimization](https://imgs.xkcd.com/comics/optimization.png)](https://xkcd.com/1691/)

PS: I appreciate the fact that Kodi is ran on many low power devices, so optimizations like this *may* make sense, but it is not unlikely that such effort is better spent elsewhere, with profiling tools to show what is spending unnecessary CPU and GPU cycles.

Reply:
@oddstr13 You are wise amongst your peers.

We struggle because as our team's collective idea of code formatting evolves, the linter can't keep up. This is good, because our team has varied valuable opinions that precede and influence our code formatting guidelines.

Soon it becomes a question of values instead of technology, which is where the "runtime optimization" fades.

Consider that *5 years* has gone into this PR. The entire team has nothing but respect for the author that is contributing to our project. We are willing to work, review, and event adapt to get this feature in.

Of course, our process refined over the last 20 years requires PR comments and change requests, both consequential and inconsequential, both easy and exhausting, and yes, many rounds of review.

Welcome to the process of open source, which with all its messiness, should see this feature successfully merged.